product_contract/models/contract_line.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2017 LasLabs Inc.
+# Copyright 2018 ACSONE SA/NV.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import api, fields, models
+
+
+class ContractLine(models.Model):
+    _inherit = "contract.line"
+    _rec_name = "display_name"
+
+    sale_order_line_id = fields.Many2one(
+        comodel_name="sale.order.line",
+        string="Sale Order Line",
+        required=False,
+        copy=False,
+    )
+
+    def _prepare_invoice_line(self):
+        res = super(ContractLine, self)._prepare_invoice_line()
+        if self.sale_order_line_id and res:
+            res["sale_line_ids"] = [(6, 0, [self.sale_order_line_id.id])]
+        return res
+
+    def _get_auto_renew_rule_type(self):
+        """monthly last day don't make sense for auto_renew_rule_type"""
+        self.ensure_one()
+        if self.recurring_rule_type == "monthlylastday":
+            return "monthly"
+        return self.recurring_rule_type
+
+    @api.onchange("product_id")
+    def _onchange_product_id_recurring_info(self):
```

Comment:
It's maybe better to avoid onchanges as fields are now computed and override compute methods

Reply:
3 of these fields have compute methods. If I wanted to override these, is it possible to extend their depends list? Like going from this:
`@api.depends("contract_id.recurring_rule_type", "contract_id.line_recurrence")`
to this:
`@api.depends("contract_id.recurring_rule_type", "contract_id.line_recurrence", "product_id")`

What about the rest of the fields? Should I leave them in the onchange method?

All suggestions are welcome @rousseldenis

Reply:
> 3 of these fields have compute methods. If I wanted to override these, is it possible to extend their depends list? Like going from this: `@api.depends("contract_id.recurring_rule_type", "contract_id.line_recurrence")` to this: `@api.depends("contract_id.recurring_rule_type", "contract_id.line_recurrence", "product_id")`


I asked Odoo for this question before and overrides could just add "product_id" as they keep previous ones.

> What about the rest of the fields? Should I leave them in the onchange method?

Yes, unfortunately :-)